pkg/apis/core/validation/validation_test.go

```diff
@@ -10552,29 +10552,417 @@ func TestValidateServiceCreate(t *testing.T) {
+			tweakSvc: func(s *core.Service) {
+				s.Spec.IPFamilyPolicy = &requireDualStack
+				s.Spec.ClusterIPs = []string{"2001::1", "garbage-ip"}
+			},
+			numErrs:              1,
+			dualStackFeatureGate: true,
+		},
+		{
+			name: "invalid, NONE + IP",
+			tweakSvc: func(s *core.Service) {
+				s.Spec.IPFamilyPolicy = &requireDualStack
+				s.Spec.ClusterIPs = []string{"None", "2001::1"}
+			},
+			numErrs:              1,
+			dualStackFeatureGate: true,
+		},
+		{
+			name: "invalid, IP + NONE",
+			tweakSvc: func(s *core.Service) {
+				s.Spec.IPFamilyPolicy = &requireDualStack
+				s.Spec.ClusterIPs = []string{"2001::1", "None"}
+			},
+			numErrs:              1,
+			dualStackFeatureGate: true,
+		},
+		{
+			name: "invalid, EMPTY STRING + IP",
+			tweakSvc: func(s *core.Service) {
+				s.Spec.IPFamilyPolicy = &requireDualStack
+				s.Spec.ClusterIPs = []string{"", "2001::1"}
+			},
+			numErrs:              1,
+			dualStackFeatureGate: true,
+		},
+		{
+			name: "invalid, IP + EMPTY STRING",
+			tweakSvc: func(s *core.Service) {
+				s.Spec.IPFamilyPolicy = &requireDualStack
+				s.Spec.ClusterIPs = []string{"2001::1", ""}
+			},
+			numErrs:              1,
+			dualStackFeatureGate: true,
+		},
+		{
+			name: "invalid, same ip family (v6)",
+			tweakSvc: func(s *core.Service) {
+				s.Spec.IPFamilyPolicy = &requireDualStack
+				s.Spec.ClusterIPs = []string{"2001::1", "2001::4"}
+				s.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol, core.IPv6Protocol}
+			},
+			numErrs:              2,
+			dualStackFeatureGate: true,
+		},
+		{
+			name: "invalid, same ip family (v4)",
+			tweakSvc: func(s *core.Service) {
+				s.Spec.IPFamilyPolicy = &requireDualStack
+				s.Spec.ClusterIPs = []string{"10.0.0.1", "10.0.0.10"}
+				s.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol, core.IPv6Protocol}
+
+			},
+			numErrs:              2,
+			dualStackFeatureGate: true,
+		},
+		{
+			name: "invalid, more than two ips",
+			tweakSvc: func(s *core.Service) {
+				s.Spec.IPFamilyPolicy = &requireDualStack
+				s.Spec.ClusterIPs = []string{"10.0.0.1", "2001::1", "10.0.0.10"}
+				s.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol, core.IPv6Protocol}
+			},
+			numErrs:              1,
+			dualStackFeatureGate: true,
+		},
+		{
+			name: "invalid, multi ip, dualstack not set",
+			tweakSvc: func(s *core.Service) {
+				s.Spec.IPFamilyPolicy = &singleStack
+				s.Spec.ClusterIPs = []string{"10.0.0.1", "2001::1"}
+				s.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol, core.IPv6Protocol}
+			},
+			numErrs:              2,
+			dualStackFeatureGate: true,
+		},
+
+		{
+			name: "invalid, multi ip, dual stack feature gate disabled",
+			tweakSvc: func(s *core.Service) {
+				s.Spec.IPFamilyPolicy = &requireDualStack
+				s.Spec.ClusterIPs = []string{"10.0.0.1", "2001::1"}
+			},
+			numErrs:              2,
+			dualStackFeatureGate: false,
+		},
+		{
+			name: "valid, headless-no-selector + multi family + gate off",
```

Comment:
and this

Reply:
I will keep this. it is a valid case (it will make more sense once gate is removed entirely from the test.)